mbed-client-pal/Source/Port/Reference-Impl/OS_Specific/ZephyrOS/RTOS/pal_plat_rtos.c

```diff
@@ -322,21 +302,22 @@ palStatus_t pal_plat_osTimerStop(palTimerID_t timerID)
     k_mutex_lock(&timer_data->lock, K_FOREVER);
 
     /* Mark timeout as stopped. */
-    timer_data->timeout = UINT64_MAX;
+    timer_data->timeout = INT64_MAX;
     if (timer_data->period) {
-        timer_data->period = UINT64_MAX;
+        timer_data->period = INT32_MAX;
     }
 
-    if (!k_work_cancel_delayable(&timer_data->work)) {
-        k_sem_give(&timer_data->busy);
-    } else {
-        /* Callback cannot be stopped.
-         * Let in-callback code to handle the restart.
+    if (timer_data->tid == k_current_get()) {
+        /* Timer stopped from its callback.
+         * Let in-callback code to handle the stop.
          */
+        k_mutex_unlock(&timer_data->lock);
+    } else {
+        /* Stop called from other other thread */
+        k_mutex_unlock(&timer_data->lock);
+        k_work_cancel_delayable_sync(&timer_data->dwork, &timer_data->work_sync);
```

Comment:
I suggest switching to normal cancel and moving under lock. Then perform cancel sync only on timer delete.
As discussed offline I think that having the cancel outside lock will lead to incoherency between work state and timer data.

I am wondering if the easiest solution would not be to simply throw away the cancel altogether. Callback will handle itself and not call upper layer if timer was stopped. In such case we would only need to call cancel sync on timer delete.

Alternatively we can introduce api lock on start and stop. Or we could rethink if there is a way to have lock removed from callback. Both of these I find a bit complicated. So what would you say about abandoning the cancel in stop?

Reply:
For the first suggestion I will give it a try.

I don't think it would be the easiest solution to simply let callback handle each stop.
In most cases cancel would remove ongoing work so having this in TimerStop is, imho, expected. If the work could not be removed then we rely on callback finishing the job.

I think cancel should stay in stop. I will put async cancel under lock (in stop) to prevent potential race and add sync in timer delete. Will test a bit and ping back if there is something to be worried about.